packages/dom-adapters/src/InlineToolAdapter/index.ts

```diff
@@ -0,0 +1,144 @@
+import type {
+  EditorJSModel,
+  InlineFragment,
+  InlineToolData,
+  InlineToolName,
+  ModelEvents,
+  TextRange
+} from '@editorjs/model';
+import {
+  EventType,
+  TextFormattedEvent,
+  TextUnformattedEvent
+} from '@editorjs/model';
+import type { CaretAdapter } from '../CaretAdapter/index.js';
+import { type IntersectType, FormattingAction } from '@editorjs/model';
+
+export interface InlineTool {
+  name: InlineToolName;
+  intersectType: IntersectType,
+  create(data?: InlineToolData): HTMLElement;
+  getAction(index: TextRange, fragments: InlineFragment[], intersectType: IntersectType, data?: InlineToolData): { action: FormattingAction; range: TextRange };
+}
+
+/**
+ * Class handles on format model events and renders inline tools
+ * Applies format to the model
+ */
+export class InlineToolAdapter {
+  /**
+   * Editor model instance
+   */
+  #model: EditorJSModel;
+
+  /**
+   * Tools, attached to the inline tool adapter
+   */
+  #tools: Map<InlineToolName, InlineTool> = new Map();
+
+  /**
+   * Caret adapter instance for the input
+   */
+  #caretAdapter: CaretAdapter;
+
+  /**
+   * @class
+   * @param model - editor model instance
+   * @param caretAdapter - caret adapter instance
+   */
+  constructor(model: EditorJSModel, caretAdapter: CaretAdapter) {
+    this.#model = model;
+    this.#caretAdapter = caretAdapter;
+
+    /**
+     * Add event listener for model changes
+     */
+    this.#model.addEventListener(EventType.Changed, (event: ModelEvents) => this.#handleModelUpdates(event));
+  }
+
+  /**
+   * Handles text format and unformat model events
+   *
+   * @param event - model change event
+   */
+  #handleModelUpdates(event: ModelEvents): void {
+    if (event instanceof TextFormattedEvent || event instanceof TextUnformattedEvent) {
+      const tool = this.#tools.get(event.detail.data.tool);
+
+      if (tool === undefined) {
+        return;
+      }
+
+      const selection = window.getSelection();
+
+      /**
+       * Render inline tool for current range
+       */
+      if (selection) {
+        const range = selection.getRangeAt(0);
+
+        const inlineElement = tool.create();
+
+        range.surroundContents(inlineElement);
```

Comment:
That would throw an exception in some cases, see https://developer.mozilla.org/en-US/docs/Web/API/Range/surroundContents